### PR TITLE
array.prototype.flatmap: support ReadonlyArray

### PR DESCRIPTION
**Pull request by mpartel**
Please fill in this template.

- [x] Use a meaningful title for the pull request. Include the name of the package modified.
- [x] Test the change in your own code. (Compile and run.)
- [x] Add or edit tests to reflect the change. (Run with `npm test`.)
- [x] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#make-a-pull-request).
- [x] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#common-mistakes).
- [x] Run `npm run lint package-name` (or `tsc` if no `tslint.json` is present).

Select one of these and delete the others:

If changing an existing definition:
- [x] Provide a URL to documentation or source code which provides context for the suggested changes: https://www.npmjs.com/package/array.prototype.flatmap
- [x] If this PR brings the type definitions up to date with a new version of the JS library, update the version number in the header. [*no changes to the JS library*]
- [x] Include [tests for your changes](https://github.com/DefinitelyTyped/DefinitelyTyped#testing)
- [x] If you are making substantial changes, consider adding a `tslint.json` containing `{ "extends": "dtslint/dt.json" }`. If for reason the any rule need to be disabled, disable it for that line using `// tslint:disable-next-line [ruleName]` and not for whole package so that the need for disabling can be reviewed. [*already has one*]

### CODE REVIEW

**Comment by mpartel** [comment]
We might want the 3rd parameter to the callback to also be readonly, but that'd technically be a breaking change so I'm not making it here, unless a reviewer asks me to :slightly_smiling_face:

File: types/array.prototype.flatmap/auto.d.ts [hunk 1]
```diff
@@ -1,6 +1,13 @@
 interface Array<T> {
     flatMap<U, R extends object | undefined = undefined>(
-        fn: (this: R, x: T, index: number, array: this) => U[],
+        fn: (this: R, x: T, index: number, array: this) => ReadonlyArray<U>,
```
Comment (ljharb) [inline types/array.prototype.flatmap/auto.d.ts:3]: Does `ReadonlyArray<U>` already include `U[]`?

**Review by ljharb — APPROVED** [review]
LGTM assuming the answer to my question is yes

File: types/array.prototype.flatmap/auto.d.ts (hunk 1 above)
Comment (hallettj) [reply types/array.prototype.flatmap/auto.d.ts:3]: I tested to be sure that returning a read-write array also works. And there are a number of existing tests in `array.prototype.flatmap-tests.ts` that I think also verify that.

**Review by hallettj — APPROVED** [review]
Looks good to me!

I'd say leave the third callback parameter as it is. Weird as the idea is the implementation does allow mutations via that parameter. I'd err on the side of simplicity, and matching the behavior of the code instead of attempting to enforce read-only use.

```ts
> const flatMap = require('array.prototype.flatmap')
undefined
> xs = [1, 2, 3]
[ 1, 2, 3 ]
> flatMap(xs, (x, i, self) => { self.push(9); return [x] })
[ 1, 2, 3 ]
> xs
[ 1, 2, 3, 9, 9, 9 ]
```

File: types/array.prototype.flatmap/auto.d.ts (hunk 1 above)
Comment (mpartel) [reply types/array.prototype.flatmap/auto.d.ts:3]: Yes, I think `ReadonlyArray` is just a subset of `Array`, or something close enough to that. Anyway, a plain array is assignable to a `ReadonlyArray`.

**Comment by mpartel** [comment]
Ready to merge

**Pull request merged**